File: delete-node-in-a-bst/delete-node-in-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    #define data val
    
    TreeNode* deleteNode(TreeNode* root, int key) {
        if(root == NULL) return NULL;
        
        else if(root->data > key)
        {
            root->left = deleteNode(root->left,key);
            return root;
        }
        else if(root->data < key)
        {
            root->right = deleteNode(root->right,key);
            return root;
        }
        else
        {
            if(root->left == NULL and root->right == NULL)
            {
                delete root;
                return NULL;
            }
            else if(root->left and root->right == NULL)
            {
                TreeNode* temp = root->left;
                root->left = NULL;
                delete root;
                return temp;
            }
            else if(root->left == NULL and root->right)
            {
                TreeNode* temp = root->right;
                root->right = NULL;
                delete root;
                return temp;
            }
            else
            {
                TreeNode* minNode = root->right;
                while(minNode->left)
                {
                    minNode = minNode->left;
                }
                int rightMin = minNode->data;
                root->data = rightMin;
                root->right = deleteNode(root->right,rightMin);
                
                return root;
            }
        }
    }
};

File: delete-node-in-a-bst/delete-node-in-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode deleteNode(TreeNode root, int key) {
        if(root == null) return null;
        
        if(root.val > key)
        {
            root.left = deleteNode(root.left,key);
            return root;
        }
        else if(root.val < key)
        {
            root.right = deleteNode(root.right,key);
            return root;
        }
        else
        {
            if(root.left == null) return root.right;
            else if(root.right == null) return root.left;
            else
            {
                TreeNode node = root.right;
                while(node.left != null) node = node.left;
                int rightMin = node.val;
                root.val = rightMin;
                root.right = deleteNode(root.right,rightMin);
                
                return root;
            }
        }
    }
}









